### ISSUE TITLE
Create and book courses within Makeradmin

### ISSUE THREAD

**Issue by nitro2k01** [makerspace/makeradmin#221]
This is a **big, low priority,** long term project concerning replacing (roughly) the role that Simple Signup has today. If someone wants to start implementing it, great, but as of writing this, there are more important things to improve in Makeradmin. The issue is posted here for inspiration. The specification below is not final. Feel free to comment and discuss below.

The feature would be a way for members to create a course event which is approved by the board, and which then allows booking, optionally with a payment.

### Creating a course event

- First the board discusses the possibility of becoming a course instructor with the member. The member is approved by being assigned to a special member group in Admin.
- Members of this group are able to create a course event by adding a title, description, date and time, price, number of attendees allowed, and maybe other relevant information.
- - It might be useful to have the ability to clone an event and (typically) just add a new date and time. Or, a template that you can create an event from.
- - It would be useful to have integration to the calendar, so that event that are published, are also visible in the calendar on the website, posted on the Slack event channel etc.
- Once submitted, the event needs to be confirmed by a member of the board, or a member in a special group with event confirmation capabilities. (Maybe not necessary if course instructors are already vetted?)

### Booking and paying

- Members can then book a slot for the course and (if it's a paid course) pay for it in the web shop, just like they would with membership time or materials.
- - Care must be taken to avoid race conditions where members are charged even when a course is full, since a popular course might be sell out from the time the member opens the page to when they're about to pay.
- - A common use case currently is giving away your slot to someone else because something else got in the way. This might be a useful feature.
- The course instructor can see a list of all booked reservations when logged in to their member account, for verification.

**Comment by emanuelen5** [comment]
See #65

**Comment by GenuineAster** [comment]
I started looking into this yesterday, since our course booking flows aren't super great. As well as improving flows and making life easier and having more control over course bookings etc, I would like to improve our ability to book courses further in advance so our calendar looks more welcoming to users and possibly because having more courses booked further in the future may reduce contention on reservations. I'm just generally interested in improving this aspect of our workflows.

I agree with the suggested workflow, but will add one component to have this mesh more easily with #222: courses and events book time on the places and resources they use, so I'll add a small feature for defining physical places in the makerspace (rooms, areas and machines) which get booked by events/courses, and the machine booking system can then use the same system to avoid scheduling conflicts. Admittedly, this is not necessary to accomplish either goal, this is partially inspired by our conversation in the board of having "course days" but coordinating use of meeting rooms etc being difficult. We are starting to have more courses and events on weekends though, so I think this will be more relevant soon.

With this system we can also do things like marking machines as out-of-order, booking maintenance slots and non-course events that may render parts of the space unusable for others. We can then have a very rich calendar integration to keep members up-to-date, and have more visibility on space usability.

I quite like the idea of templating courses/events, that allows the board to approve prices etc but give course leaders free reign for scheduling.

Not committing to finishing this, but I'm at least making some progress :) Allegations of over-engineering warmly welcome.

**Comment by nitro2k01** [comment]
#221 and #222 are kind of overlapping and would ideally depend on each other, in that a course can book a room. I still filed them as two separate issues to avoid scope creep and encourage someone with limited time to start working on just one of them alone. I'd say #221 has priority. Whereas, reconsidering it, #221 might even be a slight anti-feature as a user facing feature (as opposed to an internal feature mostly used by the course booking system). The reason being that it might create the norm that rooms should/must be booked for normal use, and prevent people from coming to work on projects spontaneously. The exception being (like today) the laser cutter and maybe some other selected machines.

**Comment by GenuineAster** [comment]
> Whereas, reconsidering it, https://github.com/makerspace/makeradmin/issues/221 might even be a slight anti-feature as a user facing feature (as opposed to an internal feature mostly used by the course booking system).

I guess you mean #222 here?

I think booking makes sense for "submit and wait" machines like 3D printers and laser cutters, just so people have an idea of availability since jobs to those machines can take a very long time, but we shouldn't normalize booking everything. Having a booking system at least allows people to be polite and social if they need to use a machine for a very long time (like.. printing a 10-part 3D print in parallel where each part takes 6 hours?) where they won't necessarily even be around to supervise. We should be careful with which machines we allow users to book though, and not expose every physical part of the space as bookable for the general membership.
